includes/utils/class-amp-mutation-utils.php

```diff
@@ -0,0 +1,325 @@
+	 *
+	 * @var array.
+	 */
+	public static $removed_attributes;
+
+	/**
+	 * The nodes that the sanitizer removed.
+	 *
+	 * @var array.
+	 */
+	public static $removed_nodes;
+
+	/**
+	 * Add the actions.
+	 *
+	 * @return void.
+	 */
+	public static function init() {
+		add_action( 'rest_api_init', array( __CLASS__, 'amp_rest_validation' ) );
+		add_action( 'save_post', array( __CLASS__, 'validate_content' ), 10, 2 );
+		add_action( 'edit_form_top', array( __CLASS__, 'display_error' ) );
+	}
+
+	/**
+	 * Tracks when a sanitizer removes an attribute or node.
+	 *
+	 * @param DOMNode|DOMElement $node The node in which there was a removal.
+	 * @param string             $removal_type The removal: 'removed_attr' for an attribute, or 'removed' for a node or element.
+	 * @param string             $attr_name The name of the attribute removed (optional).
+	 * @return void.
+	 */
+	public static function track_removed( $node, $removal_type, $attr_name = null ) {
+		if ( ( self::ATTRIBUTE_REMOVED === $removal_type ) && isset( $attr_name ) ) { // phpcs:ignore WordPress.NamingConventions.ValidVariableName.NotSnakeCaseMemberVar
+			self::$removed_attributes = self::increment_count( self::$removed_attributes, $attr_name );
+		} elseif ( ( self::NODE_REMOVED === $removal_type ) && isset( $node->nodeName ) ) {
+			self::$removed_nodes = self::increment_count( self::$removed_nodes, $node->nodeName ); // phpcs:ignore WordPress.NamingConventions.ValidVariableName.NotSnakeCaseMemberVar
+		}
+	}
+
+	/**
+	 * Tracks when a sanitizer removes an attribute or node.
+	 *
+	 * @param array  $histogram The count of attributes or nodes removed.
+	 * @param string $key The attribute or node name removed.
+	 * @return array $histogram The incremented histogram.
+	 */
+	public static function increment_count( $histogram, $key ) {
+		$current_value     = isset( $histogram[ $key ] ) ? $histogram[ $key ] : 0;
+		$histogram[ $key ] = $current_value + 1;
+		return $histogram;
+	}
+
+	/**
+	 * Gets whether a node was removed in a sanitizer.
+	 *
+	 * @return boolean.
+	 */
+	public static function was_node_removed() {
+		return ! empty( self::$removed_nodes );
+	}
+
+	/**
+	 * Processes markup, to determine AMP validity.
+	 *
+	 * Passes $markup through the AMP sanitizers.
+	 * Also passes a 'mutation_callback' to keep track of stripped attributes and nodes.
+	 *
+	 * @param string $markup The markup to process.
+	 * @return void.
+	 */
+	public static function process_markup( $markup ) {
+		$args = array(
+			'content_max_width' => ! empty( $content_width ) ? $content_width : AMP_Post_Template::CONTENT_MAX_WIDTH,
+		);
+		if ( self::is_authorized() ) {
+			$args['mutation_callback'] = 'AMP_Mutation_Utils::track_removed';
+		}
+		AMP_Content_Sanitizer::sanitize( $markup, amp_get_content_sanitizers(), $args );
+	}
+
+	/**
+	 * Registers the REST API endpoint for validation.
+	 *
+	 * @return void.
+	 */
+	public static function amp_rest_validation() {
```

Comment:
This isn't currently used, though it would be helpful for Gutenberg block validation.